### PR TITLE
dictionary learning - AM algorithms

### PR DESCRIPTION
**Pull request by gao-yuan-hangzhou**
Two alternating minimization algorithms for dictionary learning - common assumptions and convergence/recovery guarantees. Yanda Chen, Yuan Gao, Yuxiang Liu, Junhui Zhang.

### CODE REVIEW

File: chapter_6/2_dictionary_learning.tex [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+
+\subsection{Alternating minimization algorithm I}
+The first is the one considered in \cite{chatterji2017alternating}, in which the authors provide theoretical guarantees of local linear convergence. On a high level, this algorithm involves a specific robust sparse least-square estimation subroutine (for updating the coefficients $x$) and simple gradient step (for updating the dictionary $A$).
+
+We first introduce the robust sparse least-square estimation subroutine, which is defined as follows. Let $\gamma, \eta, \nu > 0$ be tuning parameters and $R>0$ be an upper bound on the entry-wise maximum dictionary estimation error ($\max_{i,j} |A_{ij} - A^*_{ij}|$). Let $(\hat{\theta}, \hat{t}, \hat{u})\in \mathbb{R}^r \times \mathbb{R}_+\times \mathbb{R}_+$ be the solution to the following convex minimization problem: 
+\begin{align}
+    \min_{\theta,t,u}\left\{ \|\theta\|_1 + \lambda t + \nv u : \left\|\frac{1}{d} A^\top (y- A\theta) \right\|_\infty\leq \gamma t + R^2 u,\, \|\theta\|_2 \leq t,\, \|\theta\|_\infty\leq u \right\}. \label{eq:mu-selector}
```
Comment (all2187) [inline chapter_6/2_dictionary_learning.tex]: "\nv" doesn't compile. Please fix the typo.

File: chapter_6/2_dictionary_learning.tex [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+\caption{Alternating Minimization for Dictionary Learning}
+% \label{alg:am-for-dl}
+\label{alg:am-for-dl-with-mus-estimator}
+\begin{algorithmic}
+\STATE \textbf{Input}: Step size $\eta$, samples $\{y^k\}_{k=1}^n$, initial estimate $A^{(0)}$, number of steps $T$, thresholds $\{\tau^{(y)}\}_{t=1}^T$, initial radius $R^{(0)}$ and robust sparse estimator parameters $\{\gamma^{(t)}\}_{t=1}^T$, $\{\lambda^{(t)}\}_{t=1}^T$, $\{\nu^{(t)}\}_{t=1}^T$.\\
+
+\FOR{$t = 1, \dots, T$}
+\FOR{$k = 1, \dots, n$} 
+    
+    Compute the robust sparse estimator $w^{k, (t)} = MUS_{\gamma^{(t)}, \lambda^{(t)}, \nu^{(t)}}(y^k, A^{(t-1)}, R^{(t-1)})$. \newline
```
Comment (all2187) [inline chapter_6/2_dictionary_learning.tex]: There is a compilation error here. I think you need to remove the newlines and add \STATE before each line.

File: chapter_6/2_dictionary_learning.tex [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+
+Next, we list and discuss the assumptions in \cite{chatterji2017alternating}. The first few are for the true dictionary $A^*$. 
+\begin{itemize}
+    \item (A1) \textbf{Incoherence}: $\left| \langle A_i^*, A_j^*\rangle \right| \leq \frac{\mu}{\sqrt{d}} $ for all $i,j\in [r]$, $i\neq j$.  This and the following (A4) essentially mean that the entries in the dictionary are sufficiently different from each other. We actually need a weaker version of this assumption. The author of the algorithm choose this one for ease of exposition.
+    \item (A2) \textbf{Normalized columns}： We require the columns of $A$ to be normalized i.e. $\|A_i^*\|_2=1$ for all $i\in [r]$.
```
Comment (all2187) [inline chapter_6/2_dictionary_learning.tex]: The colon here is a weird character that causes a compilation error. Replace it with a normal colon.

File: chapter_6/2_dictionary_learning.tex [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+
+\FOR{iterations $t=0, 1, 2, ..., T-1$}
+
+\FOR{samples $i=1, 2, ..., n$}
+
+$X(t+1)_i = \arg \min_{x \in \mathbb{R}^r}||x||_1$ such that $||Y_i - A(t)x||_2 \leq \epsilon_t$\\
+
+\ENDFOR
+
+Threshold: $X(t+1) = X(t+1) * (\mathbbm{1}[X(t+1) > 9s\epsilon_t])$\\
```
Comment (all2187) [inline chapter_6/2_dictionary_learning.tex]: Again there is a compilation error here. You need to add \STATE before each line I think.

File: chapter_6/2_dictionary_learning.tex (hunk 1 above)
Comment (all2187) [reply chapter_6/2_dictionary_learning.tex]: What is \lambda?

File: chapter_6/2_dictionary_learning.tex [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+\begin{align}
+    y^i = A^* x^{i*}, \ i\in [n], \label{eq:gen-model}
+\end{align}
+the goal is to recover $A^*\in \mathbb{R}^{d\times r}$, the dictionary, and $s$-sparse vectors $x^{i*}\in \mathbb{R}^r$.  In many applications, the dictionary is often overcomplete, that is, $r \geq d$. Note that most optimization formulations of dictionary learning result in non-convex problems that have (often infinitely) many optimal solutions. This makes it difficult to analyze its optimization and generalization properties. 
+However, fixing one of $A$ and $x$, the problem reduces to linear regression involving response variable $y$ and covariates being either $A$ or $X$, for which various classical optimization methods for least-square problems can be used. This leads to various alternating minimization heuristics, which have been proven effective in practice. Meanwhile, theoretical guarantees for alternating minimization - such as guaranteed convergence to a global optimum and local rate of convergence - have also been developed.
+
+
+\section{Alternating minimization: algorithms and theoretical guarantees}
```
Comment (all2187) [inline chapter_6/2_dictionary_learning.tex]: I don't think you need to make this a new section. It's still part of dictionary learning.

**Review by all2187 — CHANGES_REQUESTED** [review]
Excellent job! Just fix a few compilation errors and typos. Since this shouldn't take long try to make the requested changes in the next few days (ideally by Monday morning). Let me know as soon as you have made the updates. If possible, also acknowledge that you have seen this and are working on it by replying a comment on this PR.

File: chapter_6/2_dictionary_learning.tex [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+
+\subsection{Alternating minimization algorithm I}
+The first is the one considered in \cite{chatterji2017alternating}, in which the authors provide theoretical guarantees of local linear convergence. On a high level, this algorithm involves a specific robust sparse least-square estimation subroutine (for updating the coefficients $x$) and simple gradient step (for updating the dictionary $A$).
+
+We first introduce the robust sparse least-square estimation subroutine, which is defined as follows. Let $\gamma, \eta, \nu > 0$ be tuning parameters and $R>0$ be an upper bound on the entry-wise maximum dictionary estimation error ($\max_{i,j} |A_{ij} - A^*_{ij}|$). Let $(\hat{\theta}, \hat{t}, \hat{u})\in \mathbb{R}^r \times \mathbb{R}_+\times \mathbb{R}_+$ be the solution to the following convex minimization problem:
```
Comment (all2187) [inline chapter_6/2_dictionary_learning.tex]: I think \eta is supposed to be \lambda here

File: chapter_6/2_dictionary_learning.tex (hunk 1 above)
Comment (gao-yuan-hangzhou) [reply chapter_6/2_dictionary_learning.tex]: Thanks! \eta --> \lambda. The sparse robust estimation procedure is governed by (γ, λ, ν) instead.

File: chapter_6/2_dictionary_learning.tex (hunk 3 above)
Comment (gao-yuan-hangzhou) [reply chapter_6/2_dictionary_learning.tex]: Yes

File: chapter_6/2_dictionary_learning.tex (hunk 4 above)
Comment (gao-yuan-hangzhou) [reply chapter_6/2_dictionary_learning.tex]: Yes.

File: chapter_6/2_dictionary_learning.tex (hunk 5 above)
Comment (gao-yuan-hangzhou) [reply chapter_6/2_dictionary_learning.tex]: Sure.

File: chapter_6/2_dictionary_learning.tex (hunk 6 above)
Comment (gao-yuan-hangzhou) [reply chapter_6/2_dictionary_learning.tex]: Yes.

**Comment by all2187** [comment]
Great job! Thank you.

**Pull request merged**